driver-core/src/main/java/com/datastax/driver/core/querybuilder/BuiltStatement.java

```diff
@@ -235,17 +286,18 @@ public Boolean isIdempotent() {
         return !hasNonIdempotentOps();
     }
 
-    /**
-     * {@inheritDoc}
-     * <p>
-     * Note: Calling this method may trigger the underlying {@link Cluster} initialization.
-     */
     @Override
     public String toString() {
-        if (forceNoValues) {
-            return getQueryString();
+        try {
+            if (forceNoValues) {
+                return getQueryString();
+            }
+            StringBuilder queryString = buildQueryString(null, CodecRegistry.DEFAULT_INSTANCE);
+            return maybeAddSemicolon(queryString).toString();
+        } catch (CodecNotFoundException e) {
+            // Ugly but we have absolutely no context to get the registry from
+            return String.format("built query (could not generate with default codec registry: %s)", e.getMessage());
```

Comment:
> Isn't there also the issue where the value is a list containing both a non-serializable thing (e.g. function call) and a custom object?

I don't think so because that would have been caught by line 85 in `Utils.appendValue`, and the entire list would have been inlined.